### DISCUSSION TITLE
PBD solver callbacks for particle system to implement custom constraints like in FleX 'NvFlexSolverCallbackStage'

### DISCUSSION THREAD

**Discussion by mov-rax-rbx** [NVIDIA-Omniverse/PhysX#333] [Q&A]
Hi everyone.

I am currently working on a project that needs to implement custom constraint callback in particle spring system. In FleX this is achieved by using `NvFlexSolverCallbackStage` and `eNvFlexStageIterationStart` which call custom callback at each solver iteration. PhysyX also has a `PxParticleSystemCallback` and `onAdvance` method that looks like something similar to `eNvFlexStageUpdateEnd` in FleX.

Is it possible to add own callback at the start of each iteration for particle system in PhysX?

This would be great for using PhysX as collision, integration engine and implementing own custom constraint behaviour for particles using the PBD particle system.

Thank you.

**Comment by vreutskyy** [comment]
Hi @mov-rax-rbx 
Unfortunately, PhysX currently doesn't have this iteration start callback for particles system. I've created an internal feature request issue, but as we have a lot of work recently it will hardly get much attention soon.
You could try and add this callback on your own if you don't mind modifying PhysX's code. Doesn't look hard. Though, I didn't try it. Maybe there's more to it than it seems.

**Reply by mov-rax-rbx** [reply]
Thanks for the quick reply. I tried to add such a callback by myself, but unfortunately, it looks like the part of the GPU code looks closed, and I don't understand how to add a callback using an open code base. I have seen that a callback is called inside PhysXGpu_64.dll from some task through CpuDispatcher. I would be grateful if you could point me in the right direction if possible.

Thank you.

**Reply by vreutskyy** [reply]
O... You're right. Sorry, I didn't realize that part of the source code wasn't open. Yes, without it you won't be able to implement this callback by yourself. Then, the only option is the feature request I've created. Let's see if it gets approved. I'll keep you posted.
